libmscore/scale.cpp

```diff
@@ -0,0 +1,265 @@
+//
+//  This program is free software; you can redistribute it and/or modify
+//  it under the terms of the GNU General Public License version 2
+//  as published by the Free Software Foundation and appearing in
+//  the file LICENSE.GPL
+//=============================================================================
+
+#include "scale.h"
+#include "mscore.h"
+#include "note.h"
+#include <set>
+
+using namespace std;
+
+namespace Ms {
+
+//---------------------------------------------------------
+//   getNextLine
+//---------------------------------------------------------
+
+QString Scale::getNextLine(QTextStream& in)
+      {
+      while (!in.atEnd()) {
+            QString line = in.readLine();
+            if (line[0] == '!')
+                  continue;
+            return line;
+            }
+      return QString();
+      }
+
+//---------------------------------------------------------
+//   loadScale
+//---------------------------------------------------------
+
+bool Scale::loadScale(const QString& fn)
+      {
+      static set<int> supportedNbs = { 8, 12, 21, 35 };
+
+      QFile f(fn);
+      if (!f.open(QIODevice::ReadOnly)) {
+            MScore::lastError = strerror(errno);
+            return false;
+            }
+
+      QTextStream in(&f);
+
+      // Read description of the scala file
+      QString description = getNextLine(in);
+
+      int nbNotes = getNextLine(in).toInt();
+      if (supportedNbs.find(nbNotes + 1) == supportedNbs.end()) {
+            MScore::lastError = QObject::tr("MuseScore supports only Scala files with 8, 12, 21 or 35 notes");
+            return false;
+            }
+
+      for (int i = 0; i < nbNotes; ++i) {
+            QString pitchLine = getNextLine(in);
+            if (pitchLine.contains('/')) {
+                  QStringList values = pitchLine.split('/');
+                  if (values.size() != 2 || values[1].toDouble() == 0) {
+                        MScore::lastError = QObject::tr("Unsupported Scala file format.");
+                        return false;
+                        }
+
+                  originalNotes.push_back(1200 * (values[0].toDouble() / values[1].toDouble()) - 1200);
+            } else {
+                  originalNotes.push_back(pitchLine.toDouble());
+                  }
+            }
+
+      computedNotes.clear();
+      switch (originalNotes.size()) {
+            case 7:
+                  computeNotes7();
+                  break;
+            case 12:
+                  computeNotes12();
+                  break;
+            case 17:
+                  computeNotes17();
+                  break;
+            case 21:
+                  computeNotes21();
+                  break;
+            case 35:
+                  computeNotes35();
+                  break;
+            default:
+                  MScore::lastError = QObject::tr("Unsupported Scala file format.");
+                  return false;
+            }
+      return true;
+      }
+
```

Comment:
I think your suggestion is for having the keys inputted in the order of the TPC enum, instead of the Scala file format, which I did not implement. I still want to keep the Scala format. If you find there is a case for having a sharp note higher than the next bemol, you could always change the scala file and import it.